opentelemetry-api/src/opentelemetry/context/__init__.py

```diff
@@ -138,15 +138,119 @@ def __repr__(self):
         asyncio.run(main())
 """
 
+import typing
+
+from contextlib import contextmanager
+
 from .base_context import BaseRuntimeContext
 
 __all__ = ["Context"]
 
 try:
     from .async_context import AsyncRuntimeContext
 
-    Context = AsyncRuntimeContext()  # type: BaseRuntimeContext
+    _CONTEXT = AsyncRuntimeContext()  # type: BaseRuntimeContext
 except ImportError:
     from .thread_local_context import ThreadLocalRuntimeContext
 
-    Context = ThreadLocalRuntimeContext()
+    _CONTEXT = ThreadLocalRuntimeContext()
+
+
+class Context:
+    def __init__(self) -> None:
+        self.snapshot = _CONTEXT.snapshot()
+
+    def get(self, key: str) -> typing.Optional["object"]:
+        return self.snapshot.get(key)
+
+    @classmethod
+    def value(
+        cls, key: str, context: typing.Optional["Context"] = None
+    ) -> typing.Optional["object"]:
+        """
+        To access the local state of an concern, the Context API
+        provides a function which takes a context and a key as input,
+        and returns a value.
+        """
+        if context:
+            return context.get(key)
+
+        if cls.current():
+            return cls.current().get(key)
+        return None
+
+    @classmethod
+    def set_value(cls, key: str, value: "object") -> "Context":
+        """
+        To record the local state of a cross-cutting concern, the
+        Context API provides a function which takes a context, a
+        key, and a value as input, and returns an updated context
+        which contains the new value.
+
+        Args:
+            key:
+            value:
+        """
+        setattr(_CONTEXT, key, value)
+        return cls.current()
+
+    @classmethod
+    def current(cls) -> "Context":
+        """
+        To access the context associated with program execution,
+        the Context API provides a function which takes no arguments
+        and returns a Context.
+        """
+        return Context()
+
+    @classmethod
+    def set_current(cls, context: "Context") -> None:
+        """
+        To associate a context with program execution, the Context
+        API provides a function which takes a Context.
+        """
+        _CONTEXT.apply(context.snapshot)
+
+    @classmethod
+    @contextmanager
+    def use(cls, **kwargs: typing.Dict[str, object]) -> typing.Iterator[None]:
+        snapshot = {key: _CONTEXT[key] for key in kwargs}
```

Comment:
This gets a lot simpler using a single `ContextVar` to store the context:

```python
def use(cls, **kwargs: typing.Dict[str, object]) -> typing.Iterator[None]:
    token = cv.set(kwargs)
    try:
        yield
    finally:
        cv.reset(token)
```

which may be one argument for letting `AsyncRuntimeContext` implement `Context`.